File: day_09/day9.jl
#=
day9:
- Julia version: 
- Author: mike.downey
- Date: 2022-12-09
=#

import Test


function move(direction::Vector{Int}, head::Vector{Int}, tail::Vector{Int})
    new = head + direction
    return new, movetail(new, tail)
end

function movetail(head::Vector{Int}, tail::Vector{Int})
   difference = head - tail
   absolutedif = abs.(difference)
   distance = max(absolutedif...)
   if distance > 1 && absolutedif > [1,1]
       tail = tail + sign.(difference)
   elseif distance > 1
       tail = tail + [Int(ceil(e)) for e in difference/distance]
   end
   return tail
end

head = [1,3]
tail = [2,4]
step = [1,0]
head,tail = move(step, head, tail)
Test.@test tail == [2,4]
head,tail = move(step, head, tail)
Test.@test tail == [2,4]
head,tail = move(step, head, tail)
Test.@test tail == [3,3]
head,tail = move(step, head, tail)
Test.@test tail == [4,3]

function get_move(move)
    direction, step = split(move, " ")
    if direction == "U"
        vector = [0,1]
    elseif direction == "D"
        vector = [0,-1]
    elseif direction == "L"
        vector = [-1,0]
    elseif direction == "R"
        vector = [1,0]
    else
        error(move)
    end
    return vector, parse(Int, step)
end

Test.@test get_move("L 5") == ([-1,0], 5)
Test.@test get_move("R 2") == ([1,0], 2)
Test.@test get_move("D 3") == ([0,-1], 3)
Test.@test get_move("U 1") == ([0,1], 1)

data1 = readlines(open("day_09/test1.txt"))
data2 = readlines(open("day_09/test2.txt"))
data3 = readlines(open("day_09/data.txt"))

function part1(data)
    head = [0,0]
    tail = [0,0]

    visited = Set()
    for line in data
        vector,count = get_move(line)
        for i in range(1, count)
            head,tail = move(vector, head, tail)
            push!(visited, tail)
        end
    end
    println(length(visited))
    return length(visited)
end

function part2(data)
    visited = Set()
    rope = [[0,0] for i in range(1,10)]
    for line in data
        vector,count = get_move(line)
        for i in range(1, count)
            rope[1] = rope[1] + vector
            for j in range(2, 10)
                rope[j] = movetail(rope[j-1], rope[j])
            end
        push!(visited, rope[end])
        end
    end
    println(length(visited))
    return length(visited)
end


# Part 1
println("Part 1")
Test.@test part1(data1) == 13
Test.@test part1(data3) == 6197

# Part 2
println("Part 2")
Test.@test part2(data2) == 36
Test.@test part2(data3) == 2562
println("done")

File: day_09/day9.py
from aoc import aoc
import numpy as np
from scipy.spatial.distance import chebyshev


def move(direction, head, tail):
    """Takes a direction as a numpy array so (1,0) is one step right"""
    new = head + direction
    return new, movetail(new, tail)


def movetail(head, tail):
    difference = head - tail
    distance = chebyshev(head, tail)
    if distance > 1 and all(abs(difference) >= [1, 1]):
        tail = tail + np.sign(difference)
    elif distance > 1:
        tail = tail + np.ceil(difference / distance)
    return tail.astype(int)


def test():
    head = np.array([1, 3])
    tail = np.array([2, 4])
    right_step = np.array([1, 0])
    head, tail = move(right_step, head, tail)
    assert all(np.equal(tail, np.array([2, 4]))), "1"
    head, tail = move(right_step, head, tail)
    assert all(np.equal(tail, np.array([2, 4]))), '2'
    head, tail = move(right_step, head, tail)
    assert all(np.equal(tail, np.array([3, 3]))), "3"
    head, tail = move(right_step, head, tail)
    assert all(np.equal(tail, np.array([4, 3]))), "4"



moves = {
    "R": np.array([1, 0]),
    "L": np.array([-1, 0]),
    "U": np.array([0, 1]),
    "D": np.array([0, -1]),
}


def part1(data):
    head = np.array([0, 0])  # (x,y)
    tail = np.array([0, 0])

    visited = set()

    for line in data:
        angle, steps = line.split(" ")
        step = moves[angle]
        for i in range(int(steps)):
            head, tail = move(step, head, tail)
            visited.add(str(tail))
    print(len(visited))
    assert len(visited) == 6197 or len(visited) == 13


def part2(data):
    visited = set()
    rope = [np.array([0, 0])] * 10

    for line in data:
        angle, steps = line.split(" ")
        step = moves[angle]
        for _ in range(int(steps)):
            rope[0] = rope[0] + step  # move the head of the rope
            for j in range(1, 10):
                rope[j] = movetail(rope[j - 1], rope[j])
            visited.add(str(rope[-1]))

    assert len(visited) == 2562 or len(visited) == 36
    return len(visited)


def run(file):
    print("Part 1")
    data = aoc.loadlines(file)
    print(part1(data))
    print("Part 2")
    print(part2(data))


if __name__ == "__main__":
    test()
    print("Test Data 1")
    part1(aoc.loadlines("test1.txt"))
    part2(aoc.loadlines("test2.txt"))

    print("Actual Data")
    run("data.txt")

    exit()
